### PR TITLE
Case 5783: Restarting Interface allows you to have Avatar Scale Too Big

### PR DESCRIPTION
**Pull request by NeetBhagat**
Fix for case 5783: Restarting Interface allows you to have Avatar Scale Too Big
https://highfidelity.fogbugz.com/f/cases/5783

**=================
Bug details
=================**
In Settings > Avatar, you can set your Avatar scale. If you pick a number > 3 (say, 14.0), the scale is capped at 3. However, there is a bug that lets you get around the maximum:
1. Go to Settings > Avatar > Avatar Tunning grid
2. Change Avatar Scale to 14.0 (any number greater than 3)
3. Click Save Changes. 
4. Restart Interface. Notice that you are much bigger than a 3.0 scale.
Fix the bug so that even if you restart interface, your Avatar scale is capped at 3.0.

**=================
How To Test
=================**

1. Open Interface window
2. Go to Settings > Avatar
3. Change Avatar Scale to any number greater than 3. [For example 15]
4. Click Save Changes. 
5. Close the interface
6. Restart Interface

**=================
Expected Result 
=================**
Avatar scale will be 3 when you restart the interface. Not the number you have set [For example 15] before closing the interface.
If you have set any number less than 3 [For example 2]. Then after restart there will be no change in Avatar scale, it remains the same. [For example 2]

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1227/HighFidelity-Beta-PR10862-5ff0bc409b8f0def2cc3084b87ae9628a473c94c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1227/HighFidelity-Beta-PR10862-5ff0bc409b8f0def2cc3084b87ae9628a473c94c.exe

**Comment by CainFoool** [comment]
Testing.

**Comment by CainFoool** [comment]
# Test Passed
Test | Result
-------- | --------
Open Interface window | Pass
Go to Settings > Avatar | Pass
Change Avatar Scale to any number greater than 3. [For example 15] | Pass
Click Save Changes. | Pass
Close the interface | Pass
Restart Interface | Pass

Avatar Scale reset back to 3 and didn't appear as size 15 in-game or in the menu.

Let me know if this needs further testing.

**Comment by zfox23** [comment]
I am concerned that this is not a fix.

The maximum scale of an avatar is set by the server operator and is, by default, `3.0`. If a domain operator were to set this scale to, say, `2.0`, the code in this PR would not prevent a griefer from getting around that limit by setting their scale to 3.0 and restarting Interface.

Thoughts?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1318/HighFidelity-Beta-PR10862-440b94173073bb4b335f687cd4fbb7f26c1cef3f.dmg

**Comment by NeetBhagat** [comment]
Hi @zfox23 
Thank you for sharing this concern. 
I am not sure I understand your concern. Can you please give me some more information or repro steps that will break PR Code ? And can you please explain expected result ?

**Comment by anshuman64** [comment]
@NeetBhagat @zfox23 This is my fault. In the "spec" of this ticket, I did know that the max size was set by the domain operator--I thought it was a system-wide limit we had set. 
Instead of hard-coding 3.0, we need to get the variable for the max size set by the domain operator. Is that possible?

**Comment by zfox23** [comment]
Exactly, @anshuman64 :) I don't know if that's possible off the top of my head - that's for @NeetBhagat to investigate.

**Comment by anshuman64** [comment]
@NeetBhagat Repro steps to break the PR code would be: 
1. Start your Sandbox
2. In your domain settings, set Avatar > Maximum Avatar Scale to _**2.0**_
3. Run Interface and go Home
4. Set your Avatar size to _**4.0**_
5. Close and restart Interface

Expected Result: Avatar size is **_2.0_** (the max size set in domain settings)
Observed Result: Avatar size is **_3.0_** (the hard-coded value that you set)

**Comment by marko8904** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1327/HighFidelity-Beta-PR10862-440b94173073bb4b335f687cd4fbb7f26c1cef3f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1327/HighFidelity-Beta-PR10862-440b94173073bb4b335f687cd4fbb7f26c1cef3f.dmg

**Comment by themelissabrown** [comment]
@NeetBhagat please let me know when you've repro'd and this is ready for testing again

**Comment by AndrewMeadows** [comment]
This might be helpful:  this method appears to try to enforce the domain avatar scale settings:  `MyAvatar::restrictScaleFromDomainSettings()`.  There are oddities about how this feature works, in particular it appears these data fields are stored on a per-AvatarData instance which makes no sense to me but, ok:

```
AvatarData::_domainMaximumScale
AvatarData::_domainMinimumScale
```

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1436/HighFidelity-Beta-PR10862-685b1c244546e4261ae15a287269c817d811b6b0.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1436/HighFidelity-Beta-PR10862-685b1c244546e4261ae15a287269c817d811b6b0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1442/HighFidelity-Beta-PR10862-24c97183679365445b736136b78f97bacd7057f8.exe

**Comment by NeetBhagat** [comment]
@anshuman64, @zfox23 we have incorporated your suggestions. 
Code will always take max size set by the domain operator as "maximum avatar scale".
for local domain, user can set it from sandbox-> settings -> Avatar > Maximum Avatar Scale.

@CainFoool,
**Steps to test.**

1. Start your Sandbox
2. In your domain settings, set Avatar > Maximum Avatar Scale to 2.0
3. Run Interface and go Home
4. Set your Avatar size to 4.0
5. Close and restart Interface

**Expected Result**: Avatar size will be 2.0 (the max size set in domain settings)

**Comment by zfox23** [comment]
Nice. Is the behavior buggy with respect to minimum scale as well?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1529/HighFidelity-Beta-PR10862-1a47383d4fe4e1f2ab97b7a79c3d050230d1f01f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1529/HighFidelity-Beta-PR10862-1a47383d4fe4e1f2ab97b7a79c3d050230d1f01f.dmg

**Comment by CainFoool** [comment]
@NeetBhagat: Will test this once the build is available.

**Comment by NeetBhagat** [comment]
Hi @zfox23,
No it is not. we have taken clamped value of avatar scale.

@CainFoool, Yes, Please test this.

**Scenario 1) Avatar scale is set more than maximum Avatar Scale**
**Steps to test.**

1. Run Interface and go Home
2. Set your Avatar size to 4.0
3. Close Interface
4. Start your Sandbox
5. In your local domain settings, set Avatar > Maximum Avatar Scale to 2.0
6. restart Interface

**Expected Result:** Avatar size will be 2.0 (the max size set in domain settings)

**Scenario 2) Avatar scale is set less than minimum Avatar Scale
Steps to test.**

1. Run Interface and go Home
2. Set your Avatar size to 0.25
3. Close Interface
4. Start your Sandbox
5. In your local domain settings, set Avatar > Minimum Avatar Scale to 0.50
6. restart Interface

**Expected Result:** Avatar size will be 0.5 (the min size set in domain settings).

**Scenario 3) Avatar scale is set in between max-min avatar scale value
Steps to test.**

1. Run Interface and go Home
2. Set your Avatar size to 2
3. Close Interface
4. Start your Sandbox
5. In your local domain settings, set Avatar > Maximum Avatar Scale to 3 and Minimum Avatar Scale to 0.25
6. restart Interface

**Expected Result:** Avatar size will be 2 [No change]

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1531/HighFidelity-Beta-PR10862-e08e9d68216316a92dee03a1e6d6c0624368d683.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1531/HighFidelity-Beta-PR10862-e08e9d68216316a92dee03a1e6d6c0624368d683.dmg

**Comment by CainFoool** [comment]
# Test Passed

## Scenario 1) Avatar scale is set more than maximum Avatar Scale

Test | Result
-------- | --------
Run Interface and go Home | Pass
Set your Avatar size to 4.0 | Pass
Close Interface | Pass
Start your Sandbox | Pass
In your local domain settings, set Avatar > Maximum Avatar Scale to 2.0 | Pass
restart Interface | Pass 
Expected Result: Avatar size will be 2.0 (the max size set in domain settings) | Pass

## Scenario 2) Avatar scale is set less than minimum Avatar Scale

Test | Result
-------- | --------
Run Interface and go Home | Pass
Set your Avatar size to 0.25 | Pass
Close Interface | Pass
Start your Sandbox | Pass
In your local domain settings, set Avatar > Minimum Avatar Scale to 0.50 | Pass
Restart Interface | Pass
Expected Result: Avatar size will be 0.5 (the min size set in domain settings). | Pass

## Scenario 3) Avatar scale is set in between max-min avatar scale value

Test | Result
-------- | --------
Run Interface and go Home | Pass
Set your Avatar size to 2 | Pass
Close Interface | Pass
Start your Sandbox | Pass
In your local domain settings, set Avatar > Maximum Avatar Scale to 3 and Minimum Avatar Scale to 0.25 | Pass
Restart Interface | Pass
Expected Result: Avatar size will be 2 [No change] | Pass

All working perfectly!

Let me know if this needs further testing.

###### Intel i5 7600K @ 4Ghz, 16GB DDR4 2133Mhz Corsair Vengeance, Asus Strix GTX 1050Ti 3GB, HTC Vive

**Comment by tapalisa** [comment]
Someone who ran this PR does not have their environment variable set correctly. Please make sure that if you are a tester you have your environment variable set. I SEE EVERYTHING.

**Comment by NeetBhagat** [comment]
@tapalisa , Is this for me or for QA guy (@CainFoool ) ?

**Comment by CainFoool** [comment]
Mine is set in the Environment Variables correctly, is yours @NeetBhagat ?

**Comment by NeetBhagat** [comment]
@CainFoool , Obviously Yes.

**Comment by themelissabrown** [comment]
test this please

**Comment by themelissabrown** [comment]
@NeetBhagat if your environment variable is set, it is someone else that has tested this PR and hasn't set it.  No need to worry about @tapalisa's comment. Thank you.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1640/HighFidelity-Beta-PR10862-e08e9d68216316a92dee03a1e6d6c0624368d683.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1640/HighFidelity-Beta-PR10862-e08e9d68216316a92dee03a1e6d6c0624368d683.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1662/HighFidelity-Beta-PR10862-11f8f5cffccb6a6409fcaf77c13bd07275780cda.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1662/HighFidelity-Beta-PR10862-11f8f5cffccb6a6409fcaf77c13bd07275780cda.exe

**Comment by themelissabrown** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1665/HighFidelity-Beta-PR10862-11f8f5cffccb6a6409fcaf77c13bd07275780cda.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1665/HighFidelity-Beta-PR10862-11f8f5cffccb6a6409fcaf77c13bd07275780cda.dmg

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -2274,17 +2271,26 @@ void MyAvatar::restrictScaleFromDomainSettings(const QJsonObject& domainSettings
...
     }
+    // Set avatar current scale
+    Settings settings;
+    settings.beginGroup("Avatar");
+    _targetScale = loadSetting(settings, "scale", 1.0f);
 
-    qCDebug(interfaceapp, "This domain requires a minimum avatar scale of %f and a maximum avatar scale of %f",
-            (double)_domainMinimumScale, (double)_domainMaximumScale);
+    qCDebug(interfaceapp, "This domain requires a minimum avatar scale of %f and a maximum avatar scale of %f. Current avatar scale is %f",
+        _domainMinimumScale, _domainMaximumScale, _targetScale);
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: The ubuntu build is failing on this qCDebug() and also the one at line 2286: a new warning about converting a double to float or something like that.  I think this is the way to fix it (using `operator<<()`):
```
qCDebug(interfaceapp) << "This domain requires a minimum avatar scale of" << _domainMinimumScale
        << "and a maximum avatar scale of" << _domainMaximumScale
        << ". Current avatar scale is" << _targetScale << ".";
```

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1741/HighFidelity-Beta-PR10862-ff51fa71990153733ca61eeed396a8ed5e08ba36.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1741/HighFidelity-Beta-PR10862-ff51fa71990153733ca61eeed396a8ed5e08ba36.exe

File: interface/src/avatar/MyAvatar.cpp (hunk 1 above)
Comment (NeetBhagat) [inline interface/src/avatar/MyAvatar.cpp]: Resolved. Thank you!

**Comment by NeetBhagat** [comment]
@AndrewMeadows , I have resolved the comments. Thank You.

**Comment by ZappoMan** [comment]
Er... I mean to say that the changes that @AndrewMeadows requested had been made.

**Pull request merged**